include/execution.hpp

```diff
@@ -2171,7 +2173,233 @@ namespace std::execution {
+        }
+
+        friend void tag_invoke(start_t, __operation& __self) noexcept try {
+          __sh_state<_Sender> *__shared_state = __self.__shared_state_.get();
+          unique_lock __lock{__shared_state->__mutex_};
+          __state_t __state = __shared_state->__state_;
+
+          if (__state == __state_t::__completed) {
+            __lock.unlock();
+            __self.__propagate_signal();
+          }
+          else {
+            __self.__on_stop_.emplace(
+                get_stop_token(get_env(__self.__recvr_)),
+                __on_stop_requested{__shared_state->__stop_source_});
+
+            if (__shared_state->__stop_source_.stop_requested()) {
+              __lock.unlock();
+              execution::set_stopped((_Receiver&&) __self.__recvr_);
+            }
+            else {
+              __shared_state->__operation_states_.push_back(&__self);
+
+              if (__state != __state_t::__started) {
+                __shared_state->__state_ = __state_t::__started;
+                __lock.unlock();
+
+                execution::start(__shared_state->__op_state2_);
+              }
+            }
+          }
+        } catch (...) {
+          execution::set_error((_Receiver&&) __self.__recvr_, current_exception());
+        }
+      };
+
+    template <class _SenderId>
+      class __sender {
+        using _Sender = __t<_SenderId>;
+        using __sh_state = __sh_state<_Sender>;
+        template <class _Receiver>
+          using __operation = __operation<_SenderId, __x<remove_cvref_t<_Receiver>>>;
+
+        _Sender __sndr_;
+        shared_ptr<__sh_state> __shared_state_;
+
+      public:
+        template <__decays_to<__sender> _Self, receiver _Receiver>
+            requires receiver_of<_Receiver, completion_signatures_of_t<_Self>>
+          friend auto tag_invoke(connect_t, _Self&& __self, _Receiver&& __recvr)
+            noexcept(is_nothrow_constructible_v<decay_t<_Receiver>, _Receiver>)
+            -> __operation<_Receiver> {
+            return __operation<_Receiver>{(_Receiver &&) __recvr,
+                                          __self.__shared_state_};
+          }
+
+        template <__sender_queries::__sender_query _Tag, class... _As>
+            requires __callable<_Tag, const _Sender&, _As...> &&
+              __none_of<_Tag,
+                        get_completion_scheduler_t<set_value_t>,
+                        get_completion_scheduler_t<set_error_t>,
+                        get_completion_scheduler_t<set_stopped_t>>
+          friend auto tag_invoke(_Tag __tag, const __sender& __self, _As&&... __as)
+            noexcept(__nothrow_callable<_Tag, const _Sender&, _As...>)
+            -> __call_result_if_t<__sender_queries::__sender_query<_Tag>, _Tag, const _Sender&, _As...> {
+            return ((_Tag&&) __tag)(__self.__sndr_, (_As&&) __as...);
+          }
+
+        template <class... _Tys>
+        using __set_value_t = completion_signatures<set_value_t(decay_t<_Tys>&...)>;
```

Comment:
Should it be sending `const T&` to avoid data-races on multiple consumers accessing result concurrently?
Or is `T&` ok here?